### PR TITLE
[TikTok] Support Sigi-type pages, etc

### LINKED ISSUE #28741: ERROR: Unable to extract data;  (TikTok.com)
I was trying to download a video from tiktok.com that was an error but it was working last night but is not working today.

This error is occurring in VDS
But it works on localhost.

### LINKED ISSUE #30174: TikTok - Possible to download artist's entire video gallery??
- [X] I'm asking a question
- [X] I've looked through the README and FAQ for similar questions
- [X] I've searched the bugtracker for similar questions including closed ones

## Question

***TIKTOK***

I know the below is *busted*, but here goes (embarrassed),... using the most recent Youtube-DL, goal is to download an entire TikTok artist's gallery of videos.

My .bat file
`"C:\Program Files (x86)\youtube-dl\youtube-dl.exe" --config-location "C:\Program Files (x86)\youtube-dl\youtube-dl.conf" https://www.tiktok.com/@aymieandgracie?lang=en`

My .conf file

`--ffmpeg-location "C:\Program Files\ffmpeg\bin\ffmpeg.exe"
--audio-format "mp3"
--format 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/mp4'
--output "X:\10Video\Podcasts\tmp (In)\12Downloadz\%(uploader)s\%(title)s%(view_count)s.%(ext)s"`

How can the .bat file be changed to allow download of the entire gallery?

Thanks in advance.

### LINKED ISSUE #30251: Tiktok not get video url sometimes
## Checklist

- [ ] I'm reporting a broken site support = `yes`
- [ ] I've verified that I'm running youtube-dl version **2021.06.06** = `yes`
- [ ] I've checked that all provided URLs are alive and playable in a browser = `yes`
- [ ] I've checked that all URLs and arguments with special characters are properly quoted or escaped = `yes`
- [ ] I've searched the bugtracker for similar issues including closed ones = `yes`

## Verbose log

```
youtube-dl -g 'https://www.tiktok.com/@aamora_3mk/video/7028702876205632773' -v
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-g', 'https://www.tiktok.com/@aamora_3mk/video/7028702876205632773', '-v']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2021.06.06
[debug] Python version 3.8.10 (CPython) - Linux-5.11.0-40-generic-x86_64-with-glibc2.29
[debug] exe versions: ffmpeg 4.2.4, ffprobe 4.2.4
[debug] Proxy map: {}
ERROR: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 815, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 836, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 534, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/tiktok.py", line 110, in _real_extract
    page_props = self._parse_json(self._search_regex(
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 1012, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
youtube_dl.utils.RegexNotFoundError: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

## Description

Sometime it's works but sometime not work!

### LINKED ISSUE #30432: TikTok cannot get URL / download a video
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
youtube-dl https://www.tiktok.com/@leanemarts/video/7043504461670632709 --verbose
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://www.tiktok.com/@leanemarts/video/7043504461670632709', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Python version 3.8.10 (CPython) - Linux-5.4.0-1038-oracle-x86_64-with-glibc2.29
[debug] exe versions: ffmpeg 4.2.4, ffprobe 4.2.4, rtmpdump 2.4
[debug] Proxy map: {}
[TikTok] Setting up session
[TikTok] 7043504461670632709: Downloading webpage
ERROR: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 815, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 836, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 534, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/tiktok.py", line 110, in _real_extract
    page_props = self._parse_json(self._search_regex(
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 1012, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
youtube_dl.utils.RegexNotFoundError: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

## Description

TikTok probably changed something because the regex can't find what it searches for.

### LINKED ISSUE #30439: TikTok unable to extract data
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
[debug] System config: []
[debug] User config: ['--no-mtime', '--merge-output-format', 'mp4', '-o', 'D:\\YouTube-dl\\downloads\\%(id)s.%(ext)s', '--restrict-filenames']
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'https://www.tiktok.com/@imchase6/video/7044750503141444911']
[debug] Encodings: locale cp1252, fs mbcs, out cp850, pref cp1252
[debug] youtube-dl version 2021.12.17
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.19041
[debug] exe versions: ffmpeg 2021-10-28-git-e84c83ef98-full_build-www.gyan.dev, ffprobe 2021-10-28-git-e84c83ef98-full_build-www.gyan.dev
[debug] Proxy map: {}
[TikTok] Setting up session
[TikTok] 7044750503141444911: Downloading webpage
ERROR: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\YoutubeDL.py", line 815, in wrapper
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\YoutubeDL.py", line 836, in __extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\common.py", line 534, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\tiktok.py", line 112, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpupik7c6w\build\youtube_dl\extractor\common.py", line 1012, in _search_regex
youtube_dl.utils.RegexNotFoundError: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

## Description

Trying to download anything from TikTok seems to result in an error (ERROR: Unable to extract data). I tried this with multiple different links, checking that they work in the browser. Also tried to download them separately, without my user configuration, and all in one big bunch, but it didn't seem to help. It would appear that TikTok has implemented a Captcha system on the site because I had to pass it a few times over the course of testing and I suspect that it might be the reason why it is not working properly.

After trying out a multiple of links that I had saved up, I looked up a new link that was popular on the site and inserted that into the command line. It resulted in a similar verbose log and didn't work. I selected one that was quite popular and didn't seem to have any people in it to protect people's identities. Here's the link to that video: https://www.tiktok.com/@karenlafleur/video/6913759091974262022

### LINKED ISSUE #30445: TikTok Error
```
~/files/Tiktok $ youtube-dl --verbose https://vm.tiktok.com/ZM8Kf5sqY/
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'https://vm.tiktok.com/ZM8Kf5sqY/']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Python version 3.10.1 (CPython) - Linux-4.14.87+-armv8l-with-libc
[debug] exe versions: ffmpeg 4.4.1, ffprobe 4.4.1
[debug] Proxy map: {}
[generic] ZM8Kf5sqY: Requesting header
[redirect] Following redirect to https://www.tiktok.com/@elektrokurzschluss/video/7042661503035526406?_d=secCgwIARCbDRjEFSACKAESPgo8Py%2BWHrKmQ8k916xwcDBjDWlzWI3uyW3d84Z2wwgZBwA6dO9mCxynY1r%2BVVrZKE4rOvdHB5S4NR7cJt0sGgA%3D&language=de&preview_pb=0&sec_user_id=MS4wLjABAAAAwkLYynP5HnlwKtmz4yBJ6cCjmTTfkDfupLpi6htOTGYDBqQw03ruYBKCSem4mD-x&share_app_id=1233&share_item_id=7042661503035526406&share_link_id=5391ca27-a864-405f-84d3-8fec6466ad88&source=h5_m&timestamp=1640954903&u_code=de7i359c99ad13&user_id=6866698620222948358&utm_campaign=client_share&utm_medium=android&utm_source=copy&_r=1
[TikTok] Setting up session
[TikTok] 7042661503035526406: Downloading webpage
ERROR: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/youtube_dl/YoutubeDL.py", line 815, in wrapper
    return func(self, *args, **kwargs)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/youtube_dl/YoutubeDL.py", line 836, in __extract_info
    ie_result = ie.extract(url)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/youtube_dl/extractor/common.py", line 534, in extract
    ie_result = self._real_extract(url)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/youtube_dl/extractor/tiktok.py", line 110, in _real_extract
    page_props = self._parse_json(self._search_regex(
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/youtube_dl/extractor/common.py", line 1012, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
youtube_dl.utils.RegexNotFoundError: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

### LINKED ISSUE #30454: Tiktok
`youtube-dl -v https://www.tiktok.com/@sweetbodymary4/video/7045707407799880965
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'https://www.tiktok.com/@sweetbodymary4/video/7045707407799880965']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Python version 2.7.18 (CPython) - Linux-5.4.0-91-generic-x86_64-with-LinuxMint-20.2-uma
[debug] exe versions: ffmpeg 4.2.4, ffprobe 4.2.4
[debug] Proxy map: {}
[TikTok] Setting up session
[TikTok] 7045707407799880965: Downloading webpage
ERROR: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 815, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 836, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 534, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/tiktok.py", line 112, in _real_extract
    webpage, 'data'), video_id)['props']['pageProps']
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 1012, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
RegexNotFoundError: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.`

### LINKED ISSUE #30470: Unable to download from TikTok
If that may help you with updating youtube-dl
Thank you so much !!!
[OnTikTok.txt](https://github.com/ytdl-org/youtube-dl/files/7815903/OnTikTok.txt)

### PR DESCRIPTION
**Pull request by dirkf**
## Please follow the guide below

---

### Before submitting a *pull request* make sure you have:
- [ ] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
Except: this PR subsumes PR #30224 whose author also affirmed this.
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

TT switched (possibly partially) its framework from NextJS to Sigi, and the persisted state JSON sent in the page changed as a result. Instead of a `<script>` element with `id` `__NEXT_DATA__`, we get one with `id` `sigi_persisted_state` and JSON with a slightly different structure.

This PR deals with both types of page format, based on PR #30224 and [this patch](https://github.com/ytdl-org/youtube-dl/issues/30251#issuecomment-976947005) which gets more metadata.

Also, extraction could fail with a timeout (Error 60 in Windows, SSLError('The read operation timed out',) in Linux) or connection reset (Error 54 in Windows) due to some weird blocking by whatever fronts TikTok's pages (Akamai, apparenty). In order to download the page for parsing, some cookie has to be sent and a way to get it is to make a previous request to the site. The extractor fetched https://www.tiktok.com/ before doing anything else. In yt-dlp, the code fetches the webpage itself twice, commenting that you get 403 otherwise. This PR copies that tactic but instead of fetching the whole page (`GET` request) it just sends a `HEAD` request; if a page is actually returned, rather than an error with a `Set-Cookie` header, it doesn't actually have to be downloaded.

Probably resolves #28741
Resolves #30251
Resolves #30432
Resolves #30439
Resolves #30445
Resolves #30454
Resolves #30470.

Finally the non-working `TikTokUserIE` has been resurrected for accessing all the videos of a specific user.

Resolves #30174.

### CODE REVIEW

**Comment by dirkf** [comment]
Patching hints, depending on your installation type (substitute PR number 30479 and file `youtube_dl/extractor/tiktok.py` as appropriate):
* https://github.com/ytdl-org/youtube-dl/pull/30184#issuecomment-990859585
* https://github.com/ytdl-org/youtube-dl/issues/29326#issuecomment-965418428
* https://github.com/ytdl-org/youtube-dl/issues/29326#issuecomment-966349844
* https://github.com/ytdl-org/youtube-dl/issues/29326#issuecomment-972929975
* https://github.com/ytdl-org/youtube-dl/issues/29326#issuecomment-981108888.

**Comment by hessijames79** [comment]
Hi!
After your patch has worked for several days, I am now encountering new problems (with the "vanilla" youtube-dl as well): #30538

Patrick

**Comment by afterdelight** [comment]
when this merge?

File: youtube_dl/extractor/tiktok.py [hunk 1]
```diff
@@ -1,24 +1,63 @@
...
+
+    @vanilla_UA_request
+    def _request_webpage(self, *args, **kwargs):
+        return super(TikTokBaseIE, self)._request_webpage(*args, **kwargs)
+
+    def _get_SIGI_STATE(self, video_id, html):
+        state = self._parse_json(
+            get_element_by_id('SIGI_STATE', html)
+            or self._search_regex(
+                r'''(?s)<script\s[^>]*?\bid\s*=\s*(?P<q>"|'|\b)sigi-persisted-data(?P=q)[^>]*>[^=]*=\s*(?P<json>{.+?})\s*(?:;[^<]+)?</script''',
```
Comment (sulyi) [inline youtube_dl/extractor/tiktok.py:51]:
```suggestion
                 r'''(?s)<script\s[^>]*?\bid\s*=\s*(?P<q>"|'|\b)(sigi-persisted-data|SIGI_STATE)(?P=q)[^>]*>[^=]*?=?\s*(?P<json>{.+?})\s*(?:;[^<]+)?</script''',
```
id has changed in html and no longer assigned to variable.

Comment (afterdelight) [reply youtube_dl/extractor/tiktok.py:51]: can @dirkf review this?

Comment (dirkf) [reply youtube_dl/extractor/tiktok.py:51]: Check the new line `get_element_by_id('SIGI_STATE', html)` above which fixes that issue.

File: youtube_dl/extractor/tiktok.py [hunk 2]
```diff
@@ -105,34 +148,75 @@ def _real_initialize(self):
...
+        return False if TikTokIE.suitable(url) else super(TikTokBaseIE, cls).suitable(url)
 
     def _real_extract(self, url):
         user_id = self._match_id(url)
+
+        webpage = self._download_webpage(url.replace('://m.', '://www.'), user_id)
+
+        page_props = self._get_SIGI_STATE(user_id, webpage)
+        user_data = try_get(page_props, lambda x: x['UserModule']['users'], dict)
+        if user_data:
```
Comment (pukkandan) [inline youtube_dl/extractor/tiktok.py:201]: It should be
```py
        if not user_data:
            raise ExtractorError(...)
        ...
```
If the extractor returns `None`, youtube-dl will just silently exit. See https://github.com/yt-dlp/yt-dlp/issues/3776#issuecomment-1155586954

**Comment by dirkf** [comment]
As observed in https://github.com/yt-dlp/yt-dlp/issues/3776#issuecomment-1155586954 the user pages are currently redirecting to a captcha more or less whatever we do wrt cookies and UAs.

In a browser with JS disabled and UA set to `Mozilla/5.0` after clearing cookies for TT, a request to a user page gets the captcha page, and then reloading with the provided cookies opens the desired page. This doesn't happen with the extractor even with a delay between the two fetches.

**Comment by someziggyman** [comment]
Tiktok downloads are still painfully slow. Is there any chance to get these fixes merged into the main build any time soon? Just curious.

**Comment by bvoq** [comment]
Looks like every issue is about this, when will this get merged?

**Comment by OwenMelbz** [comment]
Do we think this will see the light of day? :D Was hoping to be able to use it for a little fun project!

Thanks

**Comment by kashif-umair** [comment]
I think this is also outdated now. There is no `sigi_persisted_state` in the returned HTML.

**Pull request closed without merging**